Implementation of a basic lock mechanism

arbulu89 opened the pull request:
Implementation of a basic lock mechanism.

**This PR will need changes when https://github.com/trento-project/trento/pull/70 is merged**

Context behind this implementation:

Using the KV storage doesn't allow us to manage the database entries easily, so in order to update the current entries (cluster values for example), instead of updating them, 1st the values are removed and then the new ones added. So it is not an update actually, we follow a `delete - add` way.

This creates some issues when the data is read during a write, as the write first deletes the entry. We can find ourselves in a scenario where the read returns empty or partial data if a write is happening as this exact moment. It is not a that normal situation, but happens.

To avoid this scenario Consul offers some `lock` mechanism. I have chosen to use the methods from https://github.com/hashicorp/consul/blob/master/api/lock.go and the method https://github.com/hashicorp/consul/blob/master/api/lock.go#L135 specifically. I have used this one, because it already implements an active wait to acquire and release the lock. It is really useful. Otherwise, we would need to implement this. For example, the `acquire` and `release` methods from `kv.go` don't offer this option.

With this in mind, the current implementation just uses this lock. When the write is done, a lock is set in the root key value of this component (for the cluster it would be `trento/v0/clusters` for example).

**Info**. The lock doesn't forbid to change the value from a key. It just sets the key there so you can know if a key has a lock or not, and decide upon this what to do.

**Limitations**: 
- Keep in mind that the current `lock` implementation is super basic. One of its major limitations is that it locks the entry even for read operations. This means that 2 read operations cannot be executed in parallel, and they will lock each other. The time to acquire and release the key in a read operation is super short, and I couldn't find any real delay, but I just executed some couple of parallel actions, if a lot of users are reading together this might create a delay.
- The lock might delay the answer in the http requests, which is not ideal, but I didn't find any better option why now.

rtorrero commented:
@arbulu89 as @stefanotorresi suggested yesterday, wouldn't a transaction be a better fit for this rather than implementing the lock yourself in between the two delete / write operations?

stefanotorresi commented:
I understand this is a tricky problem, but to be honest don't really like having a lock on the reads.

Just as food for thoughts, an approach could be introducing an intermediate layer between the HTTP handlers and Consul's KV store that shares the result of reads asynchronously to multiple requests via some goroutine and a channel, and either gives an up-to-date value or a cached one, so that HTTP requests are never blocked on Consul reads.

I'm also wondering if is there any way we can keep the `PUT` semantic at all, and replace the content of the keys in place in case they have changed, instead of having to separately invoke `DELETE` beforehand.

Consider also that in the future we'll likely need to implement an optimistic lock mechanism on user interactions, where the actions of one user don't interfere with the ones of another.

Besides, this entire discussion kinda suggests that maybe we're approaching the limits of what Consul can do for us as a persistent structured storage.

rtorrero commented:
Maybe I'm missing something but shouldn't we have a test where we attempt to acquire a lock that is already acquired to see if it successfully locks the resource?

arbulu89 commented:
@stefanotorresi @rtorrero The transaction usage could be possible if we only apply it `write` operations, as it doesn't have the `blocking query` capability, and this will be important most probably in the future in the moment when we want to send refreshes to the page automatically. 

I could try to modify the `PutMap` method to instead of running the `puts` individually, store them in a transaction, and we could even add the `DeleteTree` there as an optional (with a flag in the call).

I will still continue using the `DELETE+PUT` way. The update on place is really difficult, as it is super difficult to remove the `zombie` entries (in the cluster example, a resource that has been removed for example).

Finally, about having the data in some cache to be used by the web interface. We could do it in the future, but I wouldn't embark on this now. I have not found any evidence that the consul interactions are really time consuming, in fact, the reads are fast.

Let's see the result with a transaction

stefanotorresi commented:
All right! Sounds like a plan.

arbulu89 commented:
@stefanotorresi @rtorrero I have found out that consul transactions have a big limitation for us: only 64 operations are allowed. And our discoveries put more than 400 entries only for the cluster for example. So I kind of discard this option.

I think that I will try to update a bit the current implementation to remove the lock acquisition for the `read` operation, this way we could have multiple `reads` without any delay. And as we only discover every 2 minutes now, the collision between a `write` and `read` is not that common, and it doesn't take that long neither (it's a matter of ms anyway).
If in the future we need to implement some caching systems, maybe it means that we reached the capabilities consul can give as you commented @stefanotorresi hehe

arbulu89 commented:
@stefanotorresi @rtorrero I have implemented a new `LockWaitReleasead` method to do an active checking of existing locks, but this doesn't acquire it, so read operations in parallel are not blocked now.

Besides that, I have implemented some UT for the Lock mechanism. For that, I have been forced to do some "strange" changes to be able to mock some methods. Those changes are here: https://github.com/trento-project/trento/pull/71/commits/370feb8d480c3540b25f9142fefdb3d52e837cb5

I don't know if this was the best solution for the tests, but at least it works, and the whole code is tested and covered.

stefanotorresi commented on internal/cluster/cluster_kvstore.go:
```diff
@@ -43,6 +49,11 @@ func (c *Cluster) Store(client consul.Client) error {
 
+	err := client.LockWaitReleasead(consul.KvClustersPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not lock the kv for clusters")
```
this message is a bit misleading: we not trying to acquire a lock here, we're just waiting for a lock to be released.

stefanotorresi commented on internal/consul/lock.go:
```diff
@@ -0,0 +1,70 @@
+)
+
+// Created to enable unit testing
+var lockGet = func(c *client, prefix string, q *consulApi.QueryOptions) (*consulApi.KVPair, *consulApi.QueryMeta, error) {
```
(nit) functions / methods should have the verb first:
```suggestion
var getLock = func(c *client, prefix string, q *consulApi.QueryOptions) (*consulApi.KVPair, *consulApi.QueryMeta, error) {
```

stefanotorresi commented on internal/consul/consul.go:
```diff
@@ -11,6 +11,8 @@ type Client interface {
 	Health() Health
 	KV() KV
+	LockTrento(prefix string) (*consulApi.Lock, error)
+	LockWaitReleasead(prefix string) error
```
I'm also unsure about this method name; maybe
```suggestion
	WaitLock(prefix string) error
```

arbulu89 replied:
Ok, no inconvenience changing to that name

stefanotorresi commented on internal/consul/lock_test.go:
```diff
@@ -0,0 +1,146 @@
+
+		return kvPair, qMeta, nil
+	} else if fakeLockGetAttemps == -1 {
+		kvPair := &consulApi.KVPair{}
+		qMeta := &consulApi.QueryMeta{}
+
+		return kvPair, qMeta, errors.New("error")
+	} else if fakeLockGetAttemps == 0 {
+		kvPair := &consulApi.KVPair{
+			Flags:   consulApi.LockFlagValue,
+			Session: "session",
+		}
+		qMeta := &consulApi.QueryMeta{
+			LastIndex: 10,
+		}
+
+		return kvPair, qMeta, nil
+	} else if fakeLockGetAttemps == 1 {
+		kvPair := &consulApi.KVPair{
+			Flags:   consulApi.LockFlagValue,
+			Session: "session",
+		}
+		qMeta := &consulApi.QueryMeta{
+			LastIndex: 11,
+		}
+
+		assert.Equal(globalT, uint64(0xa), q.WaitIndex)
+
+		return kvPair, qMeta, nil
+	} else {
+		kvPair := &consulApi.KVPair{
+			Flags:   consulApi.LockFlagValue,
+			Session: "",
+		}
+		qMeta := &consulApi.QueryMeta{}
+
+		assert.Equal(globalT, uint64(0xb), q.WaitIndex)
+
+		return kvPair, qMeta, nil
+	}
```
This looks like `switch` block.

arbulu89 replied:
Ok, I will implement it this way

stefanotorresi commented on internal/consul/lock_test.go:
```diff
@@ -0,0 +1,146 @@
+func fakeLockGet(c *client, prefix string, q *consulApi.QueryOptions) (*consulApi.KVPair, *consulApi.QueryMeta, error) {
+	defer increaseAttempts()
+
+	assert.Equal(globalT, prefix, "test")
```
I'm not clear: why are we asserting stuff inside this mock function? Also, couldn't we pass `t` as an argument to a function returning the fake function, instead of using global? Having a shared `*testing.T` is a dangerous ticking bomb!
I think we could very well avoid this assertion entirely.

arbulu89 replied:
I cannot pass the variable as argument, this would change the faked function signature, which we cannot.
I could remove the `assert` but there is not any other place to do it. I didn't want to create a full mock, which has this functionalities

stefanotorresi commented on internal/consul/lock.go:
```diff
@@ -0,0 +1,70 @@
+		WaitTime: waitTime,
+	}
+
+WAIT:
+	// Look for an existing lock, blocking until not taken
+	// Based on: https://github.com/hashicorp/consul/blob/master/api/lock.go#L181
+	pair, meta, err := lockGet(c, prefix, &q)
+	if err != nil {
+		return fmt.Errorf("failed to read lock: %v", err)
+	}
+
+	if pair != nil && pair.Flags != consulApi.LockFlagValue {
+		return consulApi.ErrLockConflict
+	}
+
+	if pair != nil && pair.Session != "" {
+		q.WaitIndex = meta.LastIndex
+		goto WAIT
+	}
```
This is very subjective, but I never liked gotos; how do you feel about using a for loop instead? I generally find those more communicative of the fact that we're going to loop and wait until something happens.

arbulu89 replied:
Ok, I just copied what consul does. I will check how I can change it

stefanotorresi commented on internal/cluster/cluster_kvstore.go:
```diff
@@ -11,7 +11,7 @@ import (
 func (c *Cluster) getKVPath() string {
 	name := c.Name()
-	kvPath := fmt.Sprintf("%s/%s", consul.KvClustersPath, name)
+	kvPath := fmt.Sprintf("%s%s", consul.KvClustersPath, name)
```
moved to #76 for a follow up

arbulu89 commented on internal/consul/consul.go:
```diff
@@ -11,6 +11,8 @@ type Client interface {
 	Catalog() Catalog
 	Health() Health
 	KV() KV
+	LockTrento(prefix string) (*consulApi.Lock, error)
```
I named it `LockTrento` because Consul already has the `Lock` function, and I didn't want to create any confusion later on anyone trying to check the Consul `Lock` function signature and functionality.

We can name it `LockCustom` or any other name, but `Lock` might be risky

rtorrero replied:
We could use something like `LockKV` or `LockKey` or anything a bit more specific to avoid confusion

dirkmueller replied:
I agree with that, locks should be as specific as possible. do we need a global lock on any manipulation of a KV entry? or only under certain paths? the name should reflect how specific the lock is imho

stefanotorresi replied:
@arbulu89 well, this is what go packages are for: the entire name of this symbol would be `consul.Lock()`, so I don't see any room for confusion. Otherwise, we would have had to use this `Trento` suffix for every other method of this interface, which we can agree would be really ugly!

stefanotorresi replied:
`LockKey(key string)` is the upstream API signature we are wrapping around, so yes, I would probably match that exactly.

arbulu89 replied:
Well, the real consul `LockKey` is a totally different thing (it just returns an not use `Lock` struct), that's why I chose other name for this method, I didn't want people reading this `LockKey`, to think that it is the "official" consul `LockKey`. My new implementtion lock performs the lock itself. So it is not a wrapping itself as the other methods.
If you all prefer `LockKey` let's go for it, but this implements a totally different thing that the official consul `LockKey` method

stefanotorresi replied:
Okay, so while upstream only creates the handle for the lock, and one then has to call `lock()` on the lock handle to acquire it, our method creates _and_ acquires the lock in one go. Is that right?

Then our method could be called `AcquireKeyLock(key string)`. Does that work for everyone?

arbulu89 replied:
Nice, I like it. But I would put it as `AcquireLockKey(key string)`

stefanotorresi replied:
One alternative is that you mimic the exact same behaviour of the upstream method, and we don't immediately acquire the lock, we just return the lock handle configured with our own custom session parameters and then explicitly call lock on it on the consumer side.
If that's not too much hassle for you, I would actually prefer this, then we can actually call it `LockKey` and mask the upstream method, because they become basically the same thing, only ours has some different pre-set from the Consul internal locks.

arbulu89 replied:
Well, I just created this function to reduce the code in the consumers, and remove duplicates. It makes UT easier (having only in one spot, otherwise, we need to do always tricky things to mock the created `lock` instance.

Besides this, I don't really feel comfortable changing the consul api code calls. If for some reason somebody would need to use the original method, it wouldn't be possible.

Let's go with `AcquireLockKey`. It is not that important decision, and it can be reworked easily in the future if we find some need.

stefanotorresi replied:
All right then!

arbulu89 commented:
@stefanotorresi I have applied all the suggested changes in: https://github.com/trento-project/trento/pull/71/commits/647425a1788af215a41226ec3a1971babaf9a237

The unique exception is the `LockTrento` name, which is still pending. I will update it once we have agreed a name.

I have "improved" the UT to use the correct mock. It is not as straightforward as I would like, as in some scenario it will require some manual operation to update the mock.

rtorrero commented on internal/consul/lock.go:
```diff
@@ -0,0 +1,70 @@
+	return l, err
+}
+
+func (c *client) LockWaitReleasead(prefix string) error {
```
I think this is a typo. Did you mean `LockWaitReleased` ?

rtorrero replied:
Disregard this; my vscode PR review feature seems to hate me.

stefanotorresi commented:
Hey Xabi, I tried to run a live environment against this branch but I spotted the following errors in the agent log each time the discovery loops tick:

```
Error while running discovery 'sap_system_discovery': Error storing a host metadata: Error storing a new value in the KV storage: Unexpected response code: 404 ()
```

```
Error while running discovery 'ha_cluster_discovery': Error storing a host metadata: Error storing a new value in the KV storage: Unexpected response code: 404 ()
```

arbulu89 commented:
> Hey Xabi, I tried to run a live environment against this branch but I spotted the following errors in the agent log each time the discovery loops tick:
> 
> ```
> Error while running discovery 'sap_system_discovery': Error storing a host metadata: Error storing a new value in the KV storage: Unexpected response code: 404 ()
> ```
> 
> ```
> Error while running discovery 'ha_cluster_discovery': Error storing a host metadata: Error storing a new value in the KV storage: Unexpected response code: 404 ()
> ```

Thank you for checking this up. I totally messed up there (I guess I did some wrong rebase or something like this).
Fixed in https://github.com/trento-project/trento/pull/71/commits/ed6940f0c1d5988307b22428c8c8bce6806e0c2c

The fix is related to: https://github.com/trento-project/trento/pull/71/commits/ed6940f0c1d5988307b22428c8c8bce6806e0c2c
But I didn't finish doing all the changes, only the ones to fix the current issue. Let's let the issue opened, maybe there is something else we will need to check
